vm/concurrent_array.go

```diff
@@ -0,0 +1,179 @@
+					}
+
+					if len(args) == 0 {
+						return t.vm.initConcurrentArrayObject([]Object{})
+					} else {
+						arg := args[0]
+						arrayArg, ok := arg.(*ArrayObject)
+
+						if !ok {
+							return t.vm.initErrorObject(errors.TypeError, sourceLine, errors.WrongArgumentTypeFormat, classes.ArrayClass, arg.Class().Name)
+						}
+
+						return t.vm.initConcurrentArrayObject(arrayArg.Elements)
+					}
+				}
+			},
+		},
+	}
+}
+
+// Instance methods -----------------------------------------------------
+func builtinConcurrentArrayInstanceMethods() []*BuiltinMethodObject {
+	methodDefinitions := []*BuiltinMethodObject{}
+
+	for methodName, requireWriteLock := range ConcurrentArrayMethodsForwardingTable {
+		methodFunction := DefineForwardedConcurrentArrayMethod(methodName, requireWriteLock)
+		methodDefinitions = append(methodDefinitions, methodFunction)
+	}
+
+	return methodDefinitions
+}
+
+// Internal functions ===================================================
+
+// Functions for initialization -----------------------------------------
+
+func (vm *VM) initConcurrentArrayObject(elements []Object) *ConcurrentArrayObject {
+	concurrent := vm.loadConstant("Concurrent", true)
+	array := concurrent.getClassConstant("Array")
+
+	return &ConcurrentArrayObject{
+		baseObj:     &baseObj{class: array},
+		InternalArray: vm.initArrayObject(elements[:]),
+	}
+}
+
+func initConcurrentArrayClass(vm *VM) {
+	concurrent := vm.loadConstant("Concurrent", true)
+	array := vm.initializeClass("Array", false)
+
+	array.setBuiltinMethods(builtinConcurrentArrayInstanceMethods(), false)
+	array.setBuiltinMethods(builtinConcurrentArrayClassMethods(), true)
+
+	concurrent.setClassConstant(array)
+}
+
+
+// Object interface functions -------------------------------------------
+
+// toJSON returns the object's name as the JSON string format
+func (cac *ConcurrentArrayObject) toJSON() string {
+	return cac.InternalArray.toJSON()
+}
+
+// toString returns the object's name as the string format
+func (cac *ConcurrentArrayObject) toString() string {
+	return cac.InternalArray.toString()
+}
+
+// Value returns the object
+func (cac *ConcurrentArrayObject) Value() interface{} {
+	return cac.InternalArray.Elements
+}
+
+// // Helper functions -----------------------------------------------------
+
+func DefineForwardedConcurrentArrayMethod(methodName string, requireWriteLock bool) *BuiltinMethodObject {
+	return &BuiltinMethodObject{
+		Name: methodName,
+		Fn: func(receiver Object, sourceLine int) builtinMethodBody {
+			return func(t *thread, args []Object, blockFrame *normalCallFrame) Object {
+				concurrentArray := receiver.(*ConcurrentArrayObject)
+
+				if requireWriteLock {
+					concurrentArray.Lock()
+				} else {
+					concurrentArray.RLock()
+				}
+
+				// for testing sendMethod(), uncomment this block, and comment the following one.
+
+				// t.sendMethod(methodName, len(args) - 1, blockFrame, sourceLine)
```

Comment:
You don't need to minus the arg count here. The minus one need to be added in `send`  because the first arg is method name, not a real argument for the method call.

Reply:
I'll document that function since it seems that we will use it frequently in the future.